File: GeeksforGeeks/Python3/Medium/Word Search/word-search.py
from itertools import product
from collections import Counter
class Solution:
	def isWordExist(self, board, word):
		#Code here
		m, n = len(board), len(board[0])
		b = Counter(board[i][j] for i, j in product(range(m), range(n)))
		w = Counter(word)
		if len(w) > len(b): 
		    return 0
		def dfs(r, c, index):
		    if index == len(word): 
		        return 1
		    if r < 0 or r == m or c < 0 or c == n or board[r][c] != word[index]: return 0
		    board[r][c] = '@'
		    if dfs(r + 1, c, index + 1): 
		        return 1
		    if dfs(r - 1, c, index + 1): 
		        return 1
		    if dfs(r, c + 1, index + 1): 
		        return 1
		    if dfs(r, c - 1, index + 1): 
		        return 1
		    board[r][c] = word[index]
		for i, j in product(range(m), range(n)):
		    if dfs(i, j, 0): 
		        return 1
		   
		return 0


#{ 
 # Driver Code Starts
if __name__ == '__main__':
	T=int(input())
	for tt in range(T):
		n, m = map(int, input().split())
		board = []
		for i in range(n):
			a = list(input().strip().split())
			b = []
			for j in range(m):
				b.append(a[j][0])
			board.append(b)
		word = input().strip()
		obj = Solution()
		ans = obj.isWordExist(board, word)
		if(ans):
			print("1")
		else:
			print("0")

# } Driver Code Ends

File: GeeksforGeeks/C++/Medium/Word Search/word-search.cpp
//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;



// } Driver Code Ends
//User function Template for C++

class Solution {
public:
    bool solve(int i, int j, vector<vector<char>>&m, string w, int len){
        if(len == w.length()){
            return true;
        }
        if(i < 0 || j < 0 || i > m.size()-1 || j > m[0].size()-1){
            return false;
        }
        if(m[i][j] != w[len]){
            return false;
        }
        m[i][j] = '+';
        bool ans = solve(i+1,j,m,w,len+1)||solve(i-1,j,m,w,len+1)||solve(i,j-1,m,w,len+1)||solve(i,j+1,m,w,len+1);
        m[i][j] = w[len];
        return ans;
    }

    bool wordSearch(vector<vector<char>>& m, string w) {
        //code here
        for(int i=0; i<m.size(); i++){
            for(int j=0; j<m[0].size(); j++){
                if(m[i][j] == w[0] && solve(i,j,m,w,0)){
                    return true;
                }
            }
        }
        return false;
    }
};

//{ Driver Code Starts.
int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        int n,m;
        cin>>n>>m;
        vector<vector<char>> mat(n,vector<char>(m));
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                cin>>mat[i][j];
            }
        }
        string str; cin>>str;
        Solution ob;
        auto ans=ob.wordSearch(mat,str);
        cout<<ans<<"\n";
    }
    return 0;
}

// } Driver Code Ends